### PR TITLE
[TS-14]: build logic for log in screen

### PR DESCRIPTION
**Pull request by zuhair5**
## jira

[TS-14](https://shop-go.atlassian.net/browse/TS-14)

Description

- Implement username and password validation
- Password should be secure entry text, should not be visible all the time
- Password must have a special character and a number and no emoji
- If the wrong password is entered, shake the text field (left-right) and change the field’s border color to red

How to Test?

- in info.plist change the storyboard name to LoginScreen

### CODE REVIEW

File: ShopGo/LoginVC.swift [hunk 1]
```diff
@@ -18,18 +19,55 @@ class LoginVC: UIViewController {
...
     }
-
-    // Configuring textfield icons    as
+    
+    @IBAction func loginAction(_ sender: Any) {
+        let password = passwordTextfield.text
+        let username = usernameTextfield.text
+        if (isValidPassword(password: password!)) {
+            print("valid password")
+        }else{
```
Comment (haroonmaqsoodch) [inline ShopGo/LoginVC.swift:32]: space before else and after else

File: ShopGo/LoginVC.swift [hunk 2]
```diff
@@ -18,18 +19,55 @@ class LoginVC: UIViewController {
...
+            passwordTextfield.isError(baseColor: UIColor.gray.cgColor, numberOfShakes: 4, revert: true)
+            redUnderlineView.isHidden = false
+            print("invalid password")
+        }
+        if (isValidUsername(username: username!)) {
+            print("valid username")
+        }else{
```
Comment (haroonmaqsoodch) [inline ShopGo/LoginVC.swift:39]: same space issue

File: ShopGo/LoginVC.swift [hunk 3]
```diff
@@ -18,18 +19,55 @@ class LoginVC: UIViewController {
...
+        self.addLeftImageToTextfield(textfield: usernameTextfield, addImage: usernameImage)
         guard let passwordImage = UIImage(named: "padlock") else {return}
-        addLeftImageToTextfield(textfield: passwordTextfield, addImage: passwordImage)
+        self.addLeftImageToTextfield(textfield: passwordTextfield, addImage: passwordImage)
+    }
+    
+    // Dismiss keyboard with return key
+    func textFieldShouldEndEditing(_ textField: UITextField) -> Bool {
+        textField.resignFirstResponder()
+        print("return")
```
Comment (haroonmaqsoodch) [inline ShopGo/LoginVC.swift:65]: remove all the print statements

File: ShopGo/LoginVC.swift [hunk 4]
```diff
@@ -18,18 +19,55 @@ class LoginVC: UIViewController {
...
+        return true
     }
 }
 
-extension LoginVC {
+extension LoginVC{
```
Comment (haroonmaqsoodch) [inline ShopGo/LoginVC.swift:70]: space before curly brace

File: ShopGo/LoginVC.swift [hunk 5]
```diff
@@ -42,4 +80,64 @@ extension LoginVC {
...
+        passwordTextfield.text = ""
+    }
+    
+    //Button enable and Disable
+    func buttonEnableDisable() {
+        if let name = usernameTextfield.text, let password = passwordTextfield.text {
+            if !name.isEmpty && !password.isEmpty {
+                loginButton.isEnabled = true
+                loginButton.alpha = 1
+            }else{
```
Comment (haroonmaqsoodch) [inline ShopGo/LoginVC.swift:96]: same

File: ShopGo/LoginVC.swift [hunk 6]
```diff
@@ -42,4 +80,64 @@ extension LoginVC {
...
+    // textfield shake
+    func isError(baseColor: CGColor, numberOfShakes shakes: Float, revert: Bool) {
+        let shake: CABasicAnimation = CABasicAnimation(keyPath: "position")
+        shake.duration = 0.07
+        shake.repeatCount = shakes
+        if revert { shake.autoreverses = true  } else { shake.autoreverses = false }
+        shake.fromValue = NSValue(cgPoint: CGPoint(x: self.center.x - 10, y: center.y))
+        shake.toValue = NSValue(cgPoint: CGPoint(x: self.center.x + 10, y: center.y))
+        layer.add(shake, forKey: "position")
+    }
```
Comment (haroonmaqsoodch) [inline ShopGo/LoginVC.swift:142]: make a separate file name it "Textfield+extension" and move this from here

**Review by haroonmaqsoodch — COMMENTED** [review]
If we type one word in the password field the login button still gets enabled even tho we need to enforce the user to have special character and number

File: ShopGo/LoginVC.swift [hunk 7]
```diff
@@ -5,8 +5,12 @@
 //
 
 import UIKit
-class LoginVC: UIViewController {
+class LoginVC: UIViewController , UITextFieldDelegate {
```
Comment (basitalich) [inline ShopGo/LoginVC.swift:9]: Conform the textfield delegate in extension and separate the code for textfield over there.

File: ShopGo/LoginVC.swift [hunk 8]
```diff
@@ -18,14 +22,48 @@ class LoginVC: UIViewController {
...
+    
+    @IBAction func loginAction(_ sender: Any) {
+        let passwordField = passwordTextfield.text
+        let usernameField = usernameTextfield.text
+        if (isValidPassword(password: passwordField!)) && passwordField == password{
+        } else {
+            passwordTextfield.isError(baseColor: UIColor.gray.cgColor, numberOfShakes: 4, revert: true)
+            redUnderlineView.isHidden = false
+        }
+        if (isValidUsername(username: usernameField!)) && usernameField == username {
```
Comment (basitalich) [inline ShopGo/LoginVC.swift:38]: Clean these conditions, they are looking weird.

**Review by basitalich — CHANGES_REQUESTED** [review]

File: ShopGo/LoginVC.swift (hunk 7 above)
Comment (zuhair5) [reply ShopGo/LoginVC.swift:9]: resolved

File: ShopGo/LoginVC.swift (hunk 8 above)
Comment (zuhair5) [reply ShopGo/LoginVC.swift]: resolved

**Review by MuhammadHuzaifa2 — APPROVED** [review]

**Pull request merged**